### DISCUSSION TITLE
WebAuthN Workflow Concerns

### DISCUSSION THREAD

**Discussion by Borvik** [npm/feedback#694] [CLI] [answered]
So I'm glad NPM is trying to use more than just TOTP, and I've started testing out the YubiKey implementation with WebAuthn.

I hope that the flow improves, because I think that the current security flow is even more cumbersome that the TOTP method.

The process of copying the URL from the notice, logging in via a web portal which generates a long OTP that must then be pasted in is very cumbersome.

To be clear - I do see a use case for that scenario (namely when you are publishing over an SSH connection).

However, when publishing from code that is local, I think the npm command should be able to directly connect to the OS and trigger the authentication and continue without having to paste a code.

Thanks for working towards this!

**Comment by MylesBorins** [comment, marked as answer]
Hey @borvik

As mentioned in our [launch post](https://github.blog/2022-05-10-enhanced-2fa-experience-for-your-npm-account/) improving the login + publish experience to make it more streamlined is being worked on right now!

The current implementation was designed for backwards compatibility and to simplify implementation for other CLIs, for example yarn support was less than 10 lines.

The streamlined process with allow you to "hit enter" to open the website to auth and not require copying a code back in the CLI. We are also exploring how to limit how often a second factor is needed for multiple publishes in a short window.

Thanks for the feedback! Your intuition is dead on, the current implementation needs a lot of improvement to be considered "done"

**Reply by Borvik** [reply]
Thanks for the response!  That does indeed sound better.

**Reply by Borvik** [reply]
As a follow-up, I found out that the Yubikey has a one-touch OTP - essentially it acts as a keyboard to enter a code which can be verified by a service (https://developers.yubico.com/Developer_Program/Guides/Touch_triggered_OTP.html).

Perhaps that could be used behind the scenes - though I'm not certain how a backend would differentiate the different OTP sources that could be used in that same manner.

Just a thought that could make Yubikey implementation almost seamless - even through remote ssh connections.
